### ISSUE TITLE
How to evaluate YOLO performance?

### ISSUE THREAD

**Issue by caisamuels** [ultralytics/ultralytics#20222]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have been using YOLOv11 for vehicle classification and would like to evaluate its performance, such as the F1 score. I have two weeks worth of classifications (147k vehicles) and nine hours of footage that could be used as the ground truth. I am new to computer vision, so I'm unsure how to evaluate it. Do I need to manually label each vehicle in the footage? What is the best way to go about this? I only have a few days left of the project, so I am quite limited by time. Thank you.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @caisamuels, thank you for reaching out and for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com/) for comprehensive information, including evaluation best practices and detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

For custom training and evaluation ❓ questions like this, please share as much information as possible, such as example images from your dataset, the structure of your ground truth data, and any logs from your evaluation runs. Make sure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to maximize your outcomes.

Join the Ultralytics community wherever you feel most comfortable! For quick real-time help, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer deeper discussions? Visit [Discourse](https://community.ultralytics.com/). Or chat on our [Subreddit](https://reddit.com/r/Ultralytics) to connect with fellow users.

## Upgrade

To ensure you aren't encountering issues already fixed, upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI runs verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help get you pointed in the right direction quickly. An Ultralytics engineer will review your question and assist you further soon.

**Comment by Y-T-G** [comment]
You need to label otherwise there's nothing to compare against for evaluation. Evaluation requires ground-truth which comes from labels.

**Comment by Petros626** [comment]
@caisamuels use this https://github.com/cvat-ai/cvat with auto-labeling.

**Comment by NEREUScode** [comment]
@caisamuels You don’t need to label all footage just annotate a sample (1 frame every few seconds) and compare it to your model’s output to calculate F1 score. Use tools like CVAT or FiftyOne to speed it up. Focus on a small but representative subset due to time limits.

**Comment by caisamuels** [comment]
> [@caisamuels](https://github.com/caisamuels) use this https://github.com/cvat-ai/cvat with auto-labeling.

Wouldn't auto-labeling defeat the point though?

**Comment by caisamuels** [comment]
> [@caisamuels](https://github.com/caisamuels) You don’t need to label all footage just annotate a sample (1 frame every few seconds) and compare it to your model’s output to calculate F1 score. Use tools like CVAT or FiftyOne to speed it up. Focus on a small but representative subset due to time limits.

What do you think would be a fair representative subset?

**Comment by Petros626** [comment]
> Wouldn't auto-labeling defeat the point though?

Which point?

From your question it's not clear for me which model type you trained. The best tested splits for train/val are 80/20 or 90/10.

**Comment by NEREUScode** [comment]
> > Wouldn't auto-labeling defeat the point though?
> 
> Which point?
> 
> From your question it's not clear for me which model type you trained. The best tested splits for train/val are 80/20 or 90/10.

A fair representative subset would include diverse scenes (different lighting, angles, and object positions) sampled across the entire footage about 5–10% of total frames, evenly spaced.

**Comment by NEREUScode** [comment]
> > [@caisamuels](https://github.com/caisamuels) You don’t need to label all footage just annotate a sample (1 frame every few seconds) and compare it to your model’s output to calculate F1 score. Use tools like CVAT or FiftyOne to speed it up. Focus on a small but representative subset due to time limits.
> 
> What do you think would be a fair representative subset?

A fair representative subset would include diverse scenes (different lighting, angles, and object positions) sampled across the entire footage about 5–10% of total frames, evenly spaced.

**Comment by glenn-jocher** [comment]
For proper model evaluation, you need labeled ground truth data to compare against your model's predictions. Given your time constraints, I recommend:

1. Sample 5-10% of your footage strategically (as suggested by @NEREUScode) - select frames that represent different conditions, times of day, and vehicle types
2. Manually label this subset using a tool like CVAT or LabelImg
3. Run your model on these same frames and compare results

Auto-labeling with another model would introduce that model's biases into your evaluation, creating a circular validation problem. For calculating metrics like F1 score, Ultralytics provides built-in validation capabilities through `model.val()` as detailed in our [Performance Metrics guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

Remember that even a small but diverse and properly labeled test set will provide more valuable insights than a larger but poorly validated one.
